api/src/main/java/com/github/eyefloaters/console/api/service/MetricsService.java

```diff
@@ -0,0 +1,125 @@
+package com.github.eyefloaters.console.api.service;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.json.JsonArray;
+import jakarta.json.JsonObject;
+import jakarta.ws.rs.WebApplicationException;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.jboss.logging.Logger;
+
+import com.github.eyefloaters.console.api.model.Metrics;
+import com.github.eyefloaters.console.api.model.Metrics.RangeEntry;
+import com.github.eyefloaters.console.api.support.PrometheusAPI;
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+@ApplicationScoped
+public class MetricsService {
+
+    public static final String METRIC_NAME = "__console_metric_name__";
+
+    @Inject
+    Logger logger;
+
+    @Inject
+    @ConfigProperty(name = "console.metrics.prometheus-url")
+    Supplier<Optional<String>> prometheusUrl;
+
+    @Inject
+    @RestClient
+    PrometheusAPI prometheusAPI;
+
+    public boolean disabled() {
+        return prometheusUrl.get().isEmpty();
+    }
+
+    CompletionStage<Map<String, List<Metrics.ValueMetric>>> queryValues(String query) {
+        String time = Double.toString(System.currentTimeMillis() / 1000d);
+
+        return fetchMetrics(
+            () -> prometheusAPI.query(query, time),
+            (metric, attributes) -> {
+                // ignore timestamp in first position
+                String value = metric.getJsonArray("value").getString(1);
+                return new Metrics.ValueMetric(value, attributes);
+            });
+    }
+
+    CompletionStage<Map<String, List<Metrics.RangeMetric>>> queryRanges(String query) {
+        Instant now = Instant.now();
+        String start = Double.toString(now.minus(30, ChronoUnit.MINUTES).toEpochMilli() / 1000d);
```

Comment:
Metrics window is currently hard-coded to 30 minutes, could be made configurable if necessary.